packages/big-design/src/components/TableNext/hooks/useSelectable/helpers.ts

```diff
@@ -0,0 +1,84 @@
+import { TableSelectable } from '../../types';
+
+export function selectParentRow<T>(
+  childrenRows: T[],
+  isExpandable: boolean,
+  isSelectedParent: boolean,
+  onSelectionChange: TableSelectable['onSelectionChange'],
+  parentRowIndex: number,
+  selectedItems: TableSelectable['selectedItems'],
+) {
+  if (isSelectedParent) {
+    delete selectedItems[parentRowIndex];
+
+    if (isExpandable) {
+      for (let index = 0; index < childrenRows.length; index++) {
+        const childRowIndex = index;
+
+        delete selectedItems[`${parentRowIndex}.${childRowIndex}`];
+      }
+    }
+
+    onSelectionChange(selectedItems);
+  } else {
+    selectedItems[parentRowIndex] = true;
+
+    if (isExpandable) {
+      for (let index = 0; index < childrenRows.length; index++) {
+        const childRowIndex = index;
+
+        selectedItems[`${parentRowIndex}.${childRowIndex}`] = true;
+      }
+    }
+
+    onSelectionChange(selectedItems);
+  }
+}
+
+export function selectChildRow(
+  childRowIndex: number,
+  isSelectedParent: boolean,
+  isTheOnlySelectedChildRow: boolean,
+  onSelectionChange: TableSelectable['onSelectionChange'],
+  parentRowIndex: number,
+  selectedItems: TableSelectable['selectedItems'],
+) {
+  if (!isSelectedParent) {
+    selectedItems[`${parentRowIndex}`] = true;
+    selectedItems[`${parentRowIndex}.${childRowIndex}`] = true;
+
+    onSelectionChange(selectedItems);
+
+    return;
+  }
+
+  const isSelectedChild = selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
+
+  if (isSelectedChild) {
+    delete selectedItems[`${parentRowIndex}.${childRowIndex}`];
+
+    if (isTheOnlySelectedChildRow) {
+      delete selectedItems[`${parentRowIndex}`];
```

Comment:
😬 Can we try and find a better way for this instead of mutating the array? Mutations are side-effects and are hard to track down if there are any issues with it. It's better to return an immutable array instead.

Reply:
I think I'm not mutating the object. Before calling this helper fn I'm creating a copy and passed the copy to the helper function. Reference: packages/big-design/src/components/TableNext/hooks/useSelectable/useSelectable.ts

I think it might be confused because I naming this prop as selectedItems, probably I should name it differently, or make the copy inside the this helper function to prevent confusion.

Reply:
You are still mutating a passed object and mutating it's contents. Could we try and make this function a bit smaller, reduce the amount of arguments and have it do one thing: take an object and return a new one. Smaller functions are highly preferred.

```ts
const selectChildRow = (selectedItems: TableSelectable['selectedItems'], childRowIndex: number, ): TableSelectable['selectedItems'] => {
  const parentRowIndex = getParentIndex(childRowIndex);
  const isSelectedParent = selectedItems[parentRowIndex] !== undefined;

  if (!isSelectedParent) {
    selectedItems[`${parentRowIndex}`] = true;
    selectedItems[`${parentRowIndex}.${childRowIndex}`] = true;

    return {
      ...selectedItems,
      [`${parentRowIndex}`]: true,
      [`${parentRowIndex}.${childRowIndex}`] = true;
    };
  }

  // ...
}

// Usage:

const newSelectedItems = selectChildRow(selectedItems, childRowIndex);

onSelectionChange(newSelectedItems);
```

Instead of deleting keys could we just update the value to false?

Reply:
This example is still mutating, left-over copy paste?